File: lib/aura/cli/shell/executor.rb
# frozen_string_literal: true

require "aura/interface/bridge"
require "aura/context"
require_relative "console_renderer"

module Aura
  module CLI
    module Shell
      class Executor
        DANGEROUS_TOOLS = %w[
          write_file
          bash_command
        ].freeze

        def initialize(project_path, runner, config_loader)
          @project_path = project_path
          # Runner is passed, but Bridge creates its own runner? 
          # Actually, we should probably wrap the passed runner or create Bridge using project_path.
          # The existing runner passed to Executor might be pre-configured.
          # But Bridge.new takes project_path.
          # Let's check where Executor is initialized. Session.rb.
          # Session creates Runner and passes it.
          # We should update Executor to use Bridge wrapping the runner, or pass runner to Bridge.
          # Let's modify Bridge to accept an optional runner.
          # For now, let's just initialize Bridge with project_path and ignore the passed runner 
          # (or better, modify Bridge to take runner).
          # To be safe and clean, let's assume we can modify Bridge to take runner.
          # But since I already wrote Bridge, I'll stick to creating a new one or I can modify Bridge.
          # Let's modify Bridge in a second step if needed. 
          # But wait, the passed runner might have state.
          # Actually, `runner` passed to Executor is `Aura::Kernel::Runner.new(project_path)`.
          # So it's fine to just use the one created by Bridge or pass it.
          # Let's modify Bridge to allow injecting runner.
          
          @bridge = Aura::Interface::Bridge.new(project_path)
          @config_loader = config_loader
          @renderer = ConsoleRenderer.new(verbose: config["verbose"])
          
          setup_bridge
        end

        def process(input, auto_mode)
          @bridge.chat(input, auto_mode: auto_mode)
        end

        private

        def config
          @config_loader.call
        end
        
        def setup_bridge
          # Register UI callbacks
          @bridge.on(:on_waiting) do |start_time, streamed_check|
            start_timer(start_time, streamed_check)
          end
          
          @bridge.on(:on_clear_waiting) do
            @timer_thread&.kill
            @renderer.on_clear_waiting
          end
          
          @bridge.on(:on_token) do |token|
            @renderer.on_token(token)
          end
          
          @bridge.on(:on_stream_end) do
            @renderer.on_stream_end
          end
          
          @bridge.on(:on_tool_start) do |tool, summary, args|
            @renderer.on_tool_start(tool, summary, args)
          end
          
          @bridge.on(:on_tool_executing) do
            @renderer.on_tool_executing
          end
          
          @bridge.on(:on_tool_result) do |result|
            @renderer.on_tool_result(result)
          end
          
          @bridge.on(:on_warning) do |msg|
            @renderer.on_warning(msg)
          end
          
          @bridge.on(:on_error) do |msg|
            @renderer.on_error(msg)
          end
          
          @bridge.on(:on_thought) do |thought, elapsed|
            @renderer.on_thought(thought, elapsed)
          end
          
          @bridge.on(:ask_confirmation) do |msg|
            @renderer.ask_confirmation(msg)
          end
          
          # Register dangerous tool check
          @bridge.register_confirmation_hook(DANGEROUS_TOOLS)
        end

        def start_timer(start_time, streamed_check)
          @timer_thread = Thread.new do
            loop do
              break if streamed_check.call
              elapsed = Time.now - start_time
              @renderer.on_waiting(elapsed)
              sleep 0.5
            end
          end
        end
      end
    end
  end
end

File: src/cli/shell/executor.ts
import { Bridge } from '../../core/interface/bridge.js';
import { Runner } from '../../core/kernel/runner.js';
import { ConsoleRenderer } from './consoleRenderer.js';

export class Executor {
  public static readonly DANGEROUS_TOOLS = ['write_file', 'bash_command'];

  private projectPath: string;
  private bridge: Bridge;
  private configLoader: () => any;
  private renderer: ConsoleRenderer;
  private timerInterval?: NodeJS.Timeout;

  constructor(projectPath: string, runner: Runner, configLoader: () => any) {
    this.projectPath = projectPath;
    this.bridge = new Bridge(projectPath, { runner });
    this.configLoader = configLoader;

    const config = this.configLoader();
    this.renderer = new ConsoleRenderer({ verbose: config?.verbose });

    this.setupBridge();
  }

  public async process(input: string, autoMode: boolean): Promise<void> {
    this.killTimer();
    try {
      await this.bridge.chat(input, { auto_mode: autoMode });
    } finally {
      this.killTimer();
    }
  }

  /**
   * Run a single goal non-interactively and return the final answer text
   */
  public async processGoal(goal: string): Promise<string | null> {
    this.killTimer();
    let resultSummary: string | null = null;
    this.bridge.on('on_final_answer', (content: string) => {
      resultSummary = content;
    });

    try {
      await this.bridge.chat(goal, { auto_mode: true });
    } finally {
      this.killTimer();
    }
    return resultSummary;
  }

  private killTimer(): void {
    if (this.timerInterval) {
      clearInterval(this.timerInterval);
      this.timerInterval = undefined;
    }
  }

  private setupBridge(): void {
    this.bridge.on('on_waiting', (startTimeMs: number, streamedCheck: () => boolean) => {
      this.startTimer(startTimeMs, streamedCheck);
    });

    this.bridge.on('on_clear_waiting', () => {
      this.killTimer();
      this.renderer.onClearWaiting();
    });

    this.bridge.on('on_token', (token: string) => {
      this.renderer.onToken(token);
    });

    this.bridge.on('on_stream_end', () => {
      this.killTimer();
      this.renderer.onStreamEnd();
    });

    this.bridge.on('on_tool_start', (tool: string, summary?: string | null, args?: any) => {
      this.renderer.onToolStart(tool, summary, args);
    });

    this.bridge.on('on_tool_executing', () => {
      this.renderer.onToolExecuting();
    });

    this.bridge.on('on_tool_result', (result: any) => {
      this.renderer.onToolResult(result);
    });

    this.bridge.on('on_warning', (msg: string) => {
      this.killTimer();
      this.renderer.onWarning(msg);
    });

    this.bridge.on('on_error', (msg: string) => {
      this.killTimer();
      this.renderer.onError(msg);
    });

    this.bridge.on('on_thought', (thought: string, elapsed?: number | null) => {
      this.renderer.onThought(thought, elapsed);
    });

    this.bridge.on('ask_confirmation', async (msg: string) => {
      return await this.renderer.askConfirmation(msg);
    });

    // Register dangerous tool confirmation hook
    this.bridge.registerConfirmationHook(Executor.DANGEROUS_TOOLS);
  }

  private startTimer(startTimeMs: number, streamedCheck: () => boolean): void {
    this.killTimer();
    this.timerInterval = setInterval(() => {
      if (streamedCheck()) {
        this.killTimer();
        return;
      }
      const elapsed = (Date.now() - startTimeMs) / 1000;
      this.renderer.onWaiting(elapsed);
    }, 500);
  }
}
